### PR TITLE
samples: net: Add UDP sample

### PR DESCRIPTION
**Pull request by simensrostad**
Add UDP sample that demonstrates basic UDP communication.
The sample supports both Wi-Fi and LTE depending on the target board.

### CODE REVIEW

**Review by lemrey — CHANGES_REQUESTED** [review]
This sample represents one of the most common use-cases for nRF91 and it is extensively used as an example/benchmark for power profiling the product. It's important that it keeps showing clearly what to do on the nRF91 to achieve low power consumption in this most common use-case of sending UDP packets.

It should remain in `samples/cellular` in its current form, moving it under `samples/net` hides everything that it is trying to show.

**Comment by simensrostad** [comment]
Marking DNM for now pending a discussion on how we want to split up our samples between net and cellular.

**Comment by lemrey** [comment]
Please create a new sample in `net/`, we use the current sample in `/cellular` to showcase LTE-related features.

**Comment by simensrostad** [comment]
> Please create a new sample in `net/`, we use the current sample in `/cellular` to showcase LTE-related features.

Will do

File: samples/net/udp/prj.conf [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#
+# Copyright (c) 2023 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+# General configurations
+CONFIG_PICOLIBC=y
```
Comment (lemrey) [inline samples/net/udp/prj.conf:8]: This is the default, no need to specify.

File: samples/net/udp/prj.conf [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+CONFIG_LOG=y
+CONFIG_ASSERT=y
+CONFIG_HW_STACK_PROTECTION=y
```
Comment (lemrey) [inline samples/net/udp/prj.conf:11]: This is the default, no need to specify.

File: samples/net/udp/prj.conf [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#
+# Copyright (c) 2023 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+# General configurations
+CONFIG_PICOLIBC=y
+CONFIG_LOG=y
+CONFIG_ASSERT=y
```
Comment (lemrey) [inline samples/net/udp/prj.conf:10]: I think we can keep asserts disabled by default in a sample

Comment (simensrostad) [reply samples/net/udp/prj.conf]: No objections, removed for now.

**Comment by simensrostad** [comment]
@lemrey @umapraseeda Comments addressed

**Review by divipillai — APPROVED** [review]

File: samples/net/udp/sample.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    integration_platforms:
+      - nrf9160dk_nrf9160_ns
+      - nrf9161dk_nrf9161_ns
+      - thingy91_nrf9160_ns
+      - nrf7002dk_nrf5340_cpuapp
+    platform_allow:
+      - nrf9160dk_nrf9160_ns
+      - nrf9161dk_nrf9161_ns
+      - thingy91_nrf9160_ns
+      - nrf7002dk_nrf5340_cpuapp
```
Comment (rlubos) [inline samples/net/udp/sample.yaml:15]: Shall we build for qemu as well since we provide a config for it?

File: samples/net/udp/src/main.c [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+	net_mgmt_init_event_callback(&l4_cb, l4_event_handler, L4_EVENT_MASK);
+	net_mgmt_add_event_callback(&l4_cb);
+
+	/* Setup handler for Zephyr NET Connection Manager Connectivity layer. */
+	net_mgmt_init_event_callback(&conn_cb, connectivity_event_handler, CONN_LAYER_EVENT_MASK);
+	net_mgmt_add_event_callback(&conn_cb);
+
+	err = server_init();
+	if (err) {
+		LOG_INF("Not able to initialize UDP server connection");
```
Comment (rlubos) [inline samples/net/udp/src/main.c:174]: The log message is a bit confusing, we're not setting up any connection here, just configuring server address.

File: samples/net/udp/src/main.c [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+	return 0;
+}
+
+static void server_disconnect(void)
+{
+	(void)close(client_fd);
+}
+
+static int server_connect(void)
```
Comment (rlubos) [inline samples/net/udp/src/main.c:82]: I'd personally use `client_start/stop()` instead of `server_connect/disconnect()`, we're not really establishing any connection with UDP. But I don't mind much if you want to keep it as is.

**Review by rlubos — COMMENTED** [review]
Just some nitpicking

File: samples/net/udp/sample.yaml (hunk 4 above)
Comment (simensrostad) [reply samples/net/udp/sample.yaml:15]: We can do that, don't see why not.

File: samples/net/udp/src/main.c (hunk 5 above)
Comment (simensrostad) [reply samples/net/udp/src/main.c:174]: Agree, I think initially the intention was to keep the nomenclature close to the socket operations that is performed

File: samples/net/udp/src/main.c (hunk 6 above)
Comment (simensrostad) [reply samples/net/udp/src/main.c:82]: Ill fix this

File: samples/net/udp/sample.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    tags: ci_build
+    extra_configs:
+      - CONFIG_WIFI_CREDENTIALS_STATIC_SSID="ssid"
```
Comment (balaji-nordic) [inline samples/net/udp/sample.yaml:18]: These configs should only be used for a configuration that is built for wifi. So I think its better to create separate configurations for LTE and WIFI

File: samples/net/udp/boards/thingy91_nrf9160_ns.overlay
Comment (balaji-nordic) [inline samples/net/udp/boards/thingy91_nrf9160_ns.overlay:1]: I am wondering if we need to create this file. Cant we point people to use the `snippets/nrf91-modem-trace-uart` when they want to have modem traces enabled? That would mean that this sample will not have traces enabled by default (which inturn involves a readme update)

File: samples/net/udp/boards/thingy91_nrf9160_ns.overlay [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+/* Enable uart1 for tracing. */
+&uart1 {
+	status = "okay";
+	current-speed = < 1000000 >;
+	hw-flow-control;
```
Comment (balaji-nordic) [inline samples/net/udp/boards/thingy91_nrf9160_ns.overlay:11]: This property should not be here (in line with what we have in `snippets/nrf91-modem-trace-uart` )

File: samples/net/udp/src/main.c [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+	if (err < 0) {
+		LOG_INF("connect, error: %d", errno);
+		err = -errno;
+		goto error;
+	}
+
+	return 0;
+
+error:
+	server_disconnect();
```
Comment (balaji-nordic) [inline samples/net/udp/src/main.c:103]: This is not needed for any of the error conditions above. So you can remove this and replace the gotos with return statements

File: samples/net/udp/src/main.c [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+	int err;
+
+	err = server_connect();
+	if (err) {
+		LOG_INF("Not able to connect to UDP server");
+		FATAL_ERROR();
+		return;
+	}
+
+	(void)k_work_reschedule(&server_transmission_work, K_NO_WAIT);
```
Comment (balaji-nordic) [inline samples/net/udp/src/main.c:118]: Would it be better to avoid ignoring return codes and log errors instead?

File: samples/net/udp/src/main.c [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+	int err;
+
+	LOG_INF("UDP sample has started");
+
+	/* Setup handler for Zephyr NET Connection Manager events. */
+	net_mgmt_init_event_callback(&l4_cb, l4_event_handler, L4_EVENT_MASK);
+	net_mgmt_add_event_callback(&l4_cb);
+
+	/* Setup handler for Zephyr NET Connection Manager Connectivity layer. */
+	net_mgmt_init_event_callback(&conn_cb, connectivity_event_handler, CONN_LAYER_EVENT_MASK);
```
Comment (balaji-nordic) [inline samples/net/udp/src/main.c:169]: Question: Why is this needed?

**Review by balaji-nordic — CHANGES_REQUESTED** [review]

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 12]
```diff
@@ -528,6 +528,8 @@ NFC samples
 Networking samples
 ------------------
 
+* Added a new :ref:`udp_sample` sample that is compatible with both LTE and Wi-Fi.
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:531]: The old sample will likely be renamed. Do you still feel this comment is necessary given the fact?

File: samples/net/udp/README.rst [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+      [00:00:00.268,920] <inf> udp_sample: UDP sample has started
+      [00:00:00.268,951] <inf> udp_sample: Bringing network interface up and connecting to the network
+      [00:00:05.557,800] <inf> udp_sample: Network connectivity established
+      [00:00:05.558,715] <inf> udp_sample: Transmitting UDP/IP payload of 38 bytes to the IP address 8.8.8.8, port number 2469
+
+Troubleshooting
+===============
+
+If you have issues with connectivity on nRF91 Series devices, see the `Cellular Monitor`_ documentation to learn how to capture modem traces in order to debug network traffic in Wireshark.
+This sample enables modem traces by default.
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply samples/net/udp/README.rst:98]: We don't really have a good way of doing this ATM other than the usual suspects for conventional debugging. Nice improvement would be do offer a a guide to our users on how to setup wifi tracing or something equivalent. cc @ioannisg

File: samples/net/udp/sample.yaml (hunk 7 above)
Comment (simensrostad) [reply samples/net/udp/sample.yaml:18]: Yeah agree, I was hoping to avoid that :smile: But ill make the change

File: samples/net/udp/boards/thingy91_nrf9160_ns.overlay
Comment (simensrostad) [reply samples/net/udp/boards/thingy91_nrf9160_ns.overlay:1]: I think we should offer modem traces enabled by default. It saves an extra step for support and is in a lot of cases needed during development anyway.

File: samples/net/udp/boards/thingy91_nrf9160_ns.overlay (hunk 8 above)
Comment (simensrostad) [reply samples/net/udp/boards/thingy91_nrf9160_ns.overlay:11]: This is copy paste from changes that @eivindj-nordic added to some of our samples. cc @eivindj-nordic

Comment (eivindj-nordic) [reply samples/net/udp/boards/thingy91_nrf9160_ns.overlay:11]: This is only added to applications that should have modem tracing enabled by default. If not, the snippet should be used.

File: samples/net/udp/src/main.c (hunk 9 above)
Comment (simensrostad) [reply samples/net/udp/src/main.c:103]: Agreed.

File: samples/net/udp/src/main.c (hunk 10 above)
Comment (simensrostad) [reply samples/net/udp/src/main.c:118]: Return values for k_work... functions are typically not checked throughout the SDK. don't have a good answer on why. Ill add checks.

**Comment by simensrostad** [comment]
@rlubos @balaji-nordic Renamed some functions and added some more error checking.
@divipillai Added your suggested change

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (simensrostad) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst]: Added

**Review by rlubos — APPROVED** [review]

**Review by divipillai — APPROVED** [review]

**Review by balaji-nordic — APPROVED** [review]

File: samples/net/udp/boards/thingy91_nrf9160_ns.conf [hunk 14]
```diff
@@ -0,0 +1,47 @@
+#
+# Copyright (c) 2023 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+# Disable the serial interface to reduce power consumption
+CONFIG_SERIAL=n
```
Comment (lemrey) [inline samples/net/udp/boards/thingy91_nrf9160_ns.conf:8]: This is enabled further down on line 18 in this same file

**Review by lemrey — APPROVED** [review]
Looks good, just a leftover in one of the project configurations.

File: samples/net/udp/boards/thingy91_nrf9160_ns.conf (hunk 14 above)
Comment (simensrostad) [reply samples/net/udp/boards/thingy91_nrf9160_ns.conf]: Thanks, removed.
